util/topgen/merge.py

```diff
@@ -87,6 +88,29 @@ def amend_ip(top, ip):
         else:
             ip_module["available_inout_list"] = []
 
+        # param_list
+        if "param_list" in ip:
+            ip_module["param_list"] = deepcopy(ip["param_list"])
+            # Remove parameters that aren't exposed at the top level.
+            for i in ip["param_list"]:
+                if i["expose"] == "false":
+                    ip_module["param_list"].remove(i)
+                    if i["name"].lower().startswith("sec"):
+                        log.warning(
+                            ip_module["name"] + " has security-critical" +
+                            " parameter " + i["name"] + " not exposed to top")
+            # Remove descriptors and create top-level name.
+            for i in ip_module["param_list"]:
+                i.pop("desc", None)
+                name = i["name"]
+                if name.lower().startswith("sec"):
+                    name_top = "Sec" + ip_module["name"].capitalize() + name[3:]
+                else:
+                    name_top = ip_module["name"].capitalize() + name
+                i["name_top"] = name_top
+        else:
+            ip_module["param_list"] = []
```

Comment:
I wondered whether this would be better done by filtering, rather than copying everything and then amending the results. I came up with this (untested). It's a few lines shorter and a bit less "imperative". (Also, I'm not convinced we still need the `deepcopy` calls which seem to be littering the code here, although I realise that's not your design). What do you think? If you like it, you'll probably also want to hoist the definition of `mod_name` to the start of the loop and tidy up the bajillion other uses of `ip_module['name']`.
```python
        mod_name = ip_module["name"]
        mod_param_list = []
        for i in ip.get("param_list", []):
            name = i["name"]
            if i["expose"] == "false":
                if name.lower().startswith("sec"):
                    log.warning("{} has security-critical parameter {} "
                                "not exposed to top".format(mod_name, name))
            else:
                # Take the exposed parameter, removing descriptors and adding a
                # top-level name.
                mod_i = {k: deepcopy(v) for k, v in i if k != "desc"}
                mod_i["name_top"] = ("Sec" + mod_name.capitalize() + name[3:]
                                     if name.lower().startswith("sec")
                                     else mod_name.capitalize() + name)
                mod_param_list.append(mod_i)
        ip_module["param_list"] = mod_param_list
```

Reply:
Thanks @rswarbrick for the suggestion, that's much more elegant! The `deepcopy()` can be left out. I had to adjust this very line a little bit.

Reply:
i "THINK" the deepcopy may have been related to a bug @eunchan found when instantiating multiple modules of the same type.  I could be wrong though...

Reply:
I've seen a few hiccups if not use `deepcopy()` when copying from IP hjson object to top-level as the structure copied from could be modified later and instantiated into multiple instances. So, I think creating a physical hjson object for each ip instance (not module) is safer choice.

Reply:
Hm, in addition I am now getting CI failures related to that. It seems that the `param_list` in `top_earlgrey.gen.hjson` previously generated using the `deepcopy()` has a no longer a deterministic order. I'll have to investigate this on Monday.

Reply:
Ah, that's probably because hjson generates `OrderedDict` objects and my suggestion above makes just a `Dict` object for `mod_i`. My personal feeling is that `OrderedDict` is pretty much always the wrong solution to a problem, but I guess that's not really something you want to address in this PR! I guess we have to use the imperative solution after all :-(

Reply:
Switched back to using `deepcopy()`. Let's see what CI says.